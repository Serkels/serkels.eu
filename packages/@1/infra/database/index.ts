//

import { PrismaClient as PrismaClient_Base } from "#prisma/client";
import process from "node:process";

//

const prismaClientSingleton = () => {
  return new PrismaClient_Base({
    log:
      process.env["NODE_ENV"] === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });
};

const prisma = prismaClientSingleton();
export type PrismaClient = typeof prisma;
export { CategoryContext, NotificationType } from "#prisma/client";
export type { Prisma } from "#prisma/client";

export default prisma;
