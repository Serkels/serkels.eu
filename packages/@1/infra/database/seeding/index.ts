//

export * from "./category/film";
export * from "./category/masterclass";
export * from "./exchange/slap";
export * from "./exchange/the_creator";
export * from "./inbox/greetings";
export * from "./opportunity/concert_20240915";
export * from "./partner/vulfpeck";
export * from "./studient/douglas";
export * from "./studient/joedart";
export * from "./studient/johan";
