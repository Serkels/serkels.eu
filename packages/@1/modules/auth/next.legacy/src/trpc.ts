//

import type { Router } from "@1.infra/trpc";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { create_nextauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { createTRPCProxyClient, httpLink, loggerLink } from "@trpc/client";
import type { JWT } from "next-auth/jwt";
import SuperJSON from "superjson";
import { z } from "zod";

//

const ENV = z
  .object({ API_URL: z.string().url() })
  .merge(NEXTAUTH_TRPCENV)
  .parse(process.env);

//

export const trpc = createTRPCProxyClient<Router>({
  links: [
    loggerLink({
      enabled: (opts) =>
        process.env.NODE_ENV === "development" ||
        (opts.direction === "down" && opts.result instanceof Error),
    }),
    httpLink({
      url: `${ENV.API_URL}/trpc`,
      headers: async () => {
        const nexaut_header = await create_nextauth_header({
          secret: ENV.NEXTAUTH_SECRET,
          token: {
            from: "@1.modules/auth.next.legacy",
            profile: { id: "SSR", image: "", name: "", role: "ADMIN", bio: "" },
          } satisfies JWT,
          maxAge: 60,
        });
        return nexaut_header;
      },
    }),
  ],
  transformer: SuperJSON,
});
