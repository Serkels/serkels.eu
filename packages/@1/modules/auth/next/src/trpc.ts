//

import type { Router } from "@1.infra/trpc";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { create_nexauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { createTRPCProxyClient, httpLink, loggerLink } from "@trpc/client";
import type { JWT } from "next-auth/jwt";
import SuperJSON from "superjson";
import { z } from "zod";

//

const ENV = z
  .object({ API_URL: z.string().url() })
  .merge(NEXTAUTH_TRPCENV)
  .parse(process.env);

//

export const trpc = createTRPCProxyClient<Router>({
  links: [
    loggerLink({
      enabled: (opts) =>
        process.env.NODE_ENV === "development" ||
        (opts.direction === "down" && opts.result instanceof Error),
    }),
    httpLink({
      url: `${ENV.API_URL}/trpc`,
      headers: async ({}) => {
        const nexaut_header = await create_nexauth_header({
          salt: "",
          secret: ENV.AUTH_SECRET,
          token: {
            from: "@1.modules/auth.next",
            profile: { id: "SSR", image: "", name: "", role: "ADMIN" },
          } satisfies JWT,
          maxAge: 60,
        });
        return nexaut_header;
      },
    }),
  ],
  transformer: SuperJSON,
});
