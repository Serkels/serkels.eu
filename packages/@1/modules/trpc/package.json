{
  "name": "@1.modules/trpc",
  "version": "0.1.0",
  "private": true,
  "sideEffects": false,
  "type": "module",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "dev": "tsc --build --watch",
    "type-check": "tsc --build"
  },
  "keywords": [],
  "dependencies": {
    "@1.infra/database": "workspace:*",
    "@1.infra/email": "workspace:*",
    "@1.modules/profile.domain": "workspace:*",
    "@douglasduteil/nextauth...trpc.prisma": "workspace:*",
    "@trpc/server": "10.43.6",
    "superjson": "2.2.1",
    "zod": "3.22.4"
  },
  "devDependencies": {
    "@1.config/typescript": "workspace:*",
    "typescript": "5.2.2"
  }
}
