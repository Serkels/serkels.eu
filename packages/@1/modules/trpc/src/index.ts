//

import type { Profile } from "@1.modules/profile.domain";
import { create_nextauth_header as any_create_nextauth_header } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import type { ProcedureBuilder, ProcedureParams } from "@trpc/server";

//

export type { Context } from "./context";
export const create_nextauth_header = any_create_nextauth_header<{
  profile: Pick<Profile, "id" | "role">;
}>;

export {
  createCallerFactory,
  mergeRouters,
  middleware,
  procedure,
  router,
} from "./trpc";

export {
  maybe_next_auth_procedure,
  next_auth_input_token,
  next_auth_procedure,
} from "./guards";

export type inferProcedureInput<TQuery extends ProcedureBuilder<any>> =
  TQuery extends ProcedureBuilder<infer TParams>
    ? TParams extends ProcedureParams
      ? { ctx: TParams["_ctx_out"]; input: TParams["_input_out"] }
      : never
    : never;

//

export const with_previous_cursor =
  <TData>(limit: number, data: TData[]) =>
  <TCursor>(predicate: (data: TData) => TCursor) => {
    let prevCursor: TCursor | undefined = undefined;
    if (data.length > limit) {
      const prev_item = data.pop()!;
      prevCursor = predicate(prev_item);
    }

    return { data, prevCursor };
  };

export const with_next_cursor =
  <TData>(limit: number, data: TData[]) =>
  <TCursor>(predicate: (data: TData) => TCursor) => {
    let next_cursor: TCursor | undefined = undefined;
    if (data.length > limit) {
      const next_item = data.pop()!;
      next_cursor = predicate(next_item);
    }

    return { data, next_cursor };
  };
