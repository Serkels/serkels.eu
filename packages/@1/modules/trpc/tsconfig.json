{
  "$schema": "https://json.schemastore.org/tsconfig",
  "extends": "@1.config/typescript/base.json",
  "compilerOptions": {}
}
