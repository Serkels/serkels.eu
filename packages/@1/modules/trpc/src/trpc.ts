//

import type { Profile } from "@1.modules/profile.domain";
import { NEXTAUTH_TRPCENV } from "@douglasduteil/nextauth...trpc.prisma/config";
import { decode } from "@douglasduteil/nextauth...trpc.prisma/jwt";
import { verify_next_auth_token } from "@douglasduteil/nextauth...trpc.prisma/trpc";
import { TRPCError, initTRPC } from "@trpc/server";
import SuperJSON from "superjson";
import { ZodError, z } from "zod";
import type { Context } from "./context";

//
export const { router, middleware, mergeRouters, procedure } = initTRPC
  .context<Context>()
  .create({
    transformer: SuperJSON,
    errorFormatter({ shape, error }) {
      return {
        ...shape,
        data: {
          ...shape.data,
          zodError:
            error.code === "BAD_REQUEST" && error.cause instanceof ZodError
              ? error.cause.flatten()
              : null,
        },
      };
    },
  });

// export type Procedure = inferProcedureBuilderResolverOptions<typeof procedure>;
export const next_auth_procedure = procedure.use(
  verify_next_auth_token<{ profile: Profile }>(
    NEXTAUTH_TRPCENV.NEXTAUTH_SECRET,
  ),
);

// export type Next_Auth_Procedure = inferProcedureBuilderResolverOptions<
//   typeof next_auth_procedure
// >;

export const next_auth_input_token = procedure
  .input(z.object({ token: z.string() }))
  .use(async ({ input, ctx, next }) => {
    try {
      const payload = (await decode({
        token: input.token,
        secret: NEXTAUTH_TRPCENV.NEXTAUTH_SECRET,
      })) as { profile: Profile };

      if (payload) {
        return next({ ctx: { ...ctx, payload } });
      }

      throw new TRPCError({
        code: "UNAUTHORIZED",
        cause: new Error("No payload"),
      });
    } catch (cause) {
      throw new TRPCError({ code: "PARSE_ERROR", cause });
    }
  });
