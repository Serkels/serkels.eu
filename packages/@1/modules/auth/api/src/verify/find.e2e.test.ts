//

import { create_douglas_student } from "@1.infra/database/seeding";
import prisma, { empty_database, migrate } from "@1.infra/database/testing";
import { createCallerFactory } from "@1.modules/trpc";
import {
  douglas_golden_nextauth_header,
  NEXTAUTH_SECRET,
} from "@1.modules/trpc/testing";
import { beforeAll, expect, setSystemTime, test } from "bun:test";
import verify_router from "./index";

//

beforeAll(() => {
  process.env["NEXTAUTH_SECRET"] = NEXTAUTH_SECRET;
});

//

beforeAll(empty_database);
beforeAll(migrate);
beforeAll(() => {
  setSystemTime(new Date("2011-11-11"));
});

//

const nextauth_header = douglas_golden_nextauth_header;

//

test("return found user id", async () => {
  await create_douglas_student(prisma);
  const caller = createCallerFactory(verify_router);
  const trpc = caller({
    headers: { ...nextauth_header },
    prisma,
  } as any);
  const response = await trpc.verify({ email: "douglas@example.com" });
  expect(response).toMatchSnapshot();
});

test("return null when user not found", async () => {
  const caller = createCallerFactory(verify_router);
  const trpc = caller({
    headers: { ...nextauth_header },
    prisma,
  } as any);
  const response = await trpc.verify({ email: "douglas@serkels.com" });
  expect(response).toMatchSnapshot();
});
