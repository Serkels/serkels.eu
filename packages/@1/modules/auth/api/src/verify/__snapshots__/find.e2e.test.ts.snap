// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`return found user id 1`] = `
{
  "id": "douglas_user_id",
}
`;

exports[`return null when user not found 1`] = `null`;
